File: internal/tools/genui/guard_test.go
package genui

import (
	"strings"
	"testing"
)

func TestSanitize_ValidSpecPassesThrough(t *testing.T) {
	spec := map[string]any{
		"title": "Order status",
		"items": []any{
			map[string]any{"type": "text", "text": "hello"},
			map[string]any{"type": "stat", "label": "Revenue", "value": "$128,430", "tone": "up"},
		},
	}
	out, count, err := Sanitize(spec, ReadOnlyNodeTypes)
	if err != nil {
		t.Fatalf("Sanitize returned error: %v", err)
	}
	if count != 2 {
		t.Fatalf("count = %d, want 2", count)
	}
	items, ok := out["items"].([]any)
	if !ok || len(items) != 2 {
		t.Fatalf("out[items] = %#v, want 2 items", out["items"])
	}
	first := items[0].(map[string]any)
	if first["type"] != "text" || first["text"] != "hello" {
		t.Fatalf("first item mangled: %#v", first)
	}
}

func TestSanitize_MissingItemsErrors(t *testing.T) {
	_, _, err := Sanitize(map[string]any{"title": "x"}, ReadOnlyNodeTypes)
	if err == nil {
		t.Fatal("expected error for missing items")
	}
}

func TestSanitize_UnknownTypeNodeDropped(t *testing.T) {
	spec := map[string]any{
		"items": []any{
			map[string]any{"type": "text", "text": "kept"},
			map[string]any{"type": "iframe", "src": "https://evil.example"},
		},
	}
	out, count, err := Sanitize(spec, ReadOnlyNodeTypes)
	if err != nil {
		t.Fatalf("Sanitize returned error: %v", err)
	}
	if count != 1 {
		t.Fatalf("count = %d, want 1 (unknown-type node dropped)", count)
	}
	items := out["items"].([]any)
	if len(items) != 1 {
		t.Fatalf("items = %#v, want 1 item after drop", items)
	}
}

func TestSanitize_OverCapNodeCountTrimmed(t *testing.T) {
	items := make([]any, 0, MaxNodes+50)
	for i := 0; i < MaxNodes+50; i++ {
		items = append(items, map[string]any{"type": "text", "text": "x"})
	}
	spec := map[string]any{"items": items}
	out, count, err := Sanitize(spec, ReadOnlyNodeTypes)
	if err != nil {
		t.Fatalf("Sanitize returned error: %v", err)
	}
	if count != MaxNodes {
		t.Fatalf("count = %d, want %d (trimmed at cap)", count, MaxNodes)
	}
	if got := len(out["items"].([]any)); got != MaxNodes {
		t.Fatalf("len(items) = %d, want %d", got, MaxNodes)
	}
}

func TestSanitize_OverDepthNodeDropped(t *testing.T) {
	// Build a card nested 10 levels deep (MaxDepth is 8) via row->children chains.
	var build func(depth int) map[string]any
	build = func(depth int) map[string]any {
		if depth == 0 {
			return map[string]any{"type": "text", "text": "leaf"}
		}
		return map[string]any{"type": "row", "children": []any{build(depth - 1)}}
	}
	spec := map[string]any{"items": []any{build(10)}}
	out, _, err := Sanitize(spec, ReadOnlyNodeTypes)
	if err != nil {
		t.Fatalf("Sanitize returned error: %v", err)
	}
	// Walk down the surviving tree and confirm it terminates before the leaf
	// (the leaf sits at depth 11, past MaxDepth=8) — i.e. its deepest "row" has
	// no children rather than the original leaf node.
	node := out["items"].([]any)[0].(map[string]any)
	depth := 1
	for {
		children, ok := node["children"].([]any)
		if !ok || len(children) == 0 {
			break
		}
		node = children[0].(map[string]any)
		depth++
	}
	if node["type"] != "row" {
		t.Fatalf("expected the deepest surviving node to be a truncated 'row' (leaf text dropped past MaxDepth), got %#v at depth %d", node, depth)
	}
	if depth > MaxDepth {
		t.Fatalf("surviving depth = %d, want <= %d", depth, MaxDepth)
	}
}

func TestSanitize_StringFieldsClamped(t *testing.T) {
	long := make([]byte, MaxStringLen+100)
	for i := range long {
		long[i] = 'a'
	}
	spec := map[string]any{
		"items": []any{
			map[string]any{"type": "text", "text": string(long)},
		},
	}
	out, _, err := Sanitize(spec, ReadOnlyNodeTypes)
	if err != nil {
		t.Fatalf("Sanitize returned error: %v", err)
	}
	text := out["items"].([]any)[0].(map[string]any)["text"].(string)
	if len(text) != MaxStringLen {
		t.Fatalf("len(text) = %d, want %d", len(text), MaxStringLen)
	}
}

func TestSanitize_ProgressValueClamped(t *testing.T) {
	spec := map[string]any{
		"items": []any{
			map[string]any{"type": "progress", "value": 150.0},
			map[string]any{"type": "progress", "value": -20.0},
		},
	}
	out, _, err := Sanitize(spec, ReadOnlyNodeTypes)
	if err != nil {
		t.Fatalf("Sanitize returned error: %v", err)
	}
	items := out["items"].([]any)
	if v := items[0].(map[string]any)["value"]; v != 100.0 {
		t.Fatalf("value = %v, want 100", v)
	}
	if v := items[1].(map[string]any)["value"]; v != 0.0 {
		t.Fatalf("value = %v, want 0", v)
	}
}

func TestSanitize_InvalidToneDropped(t *testing.T) {
	spec := map[string]any{
		"items": []any{
			map[string]any{"type": "badge", "text": "x", "tone": "rainbow"},
		},
	}
	out, _, err := Sanitize(spec, ReadOnlyNodeTypes)
	if err != nil {
		t.Fatalf("Sanitize returned error: %v", err)
	}
	node := out["items"].([]any)[0].(map[string]any)
	if _, present := node["tone"]; present {
		t.Fatalf("expected invalid tone to be dropped, got %#v", node["tone"])
	}
}

func TestSanitize_TableRowsAndColumnsCapped(t *testing.T) {
	columns := []any{"a", "b"}
	rows := make([]any, 0, MaxTableRows+20)
	for i := 0; i < MaxTableRows+20; i++ {
		rows = append(rows, []any{"x", "y"})
	}
	spec := map[string]any{
		"items": []any{
			map[string]any{"type": "table", "columns": columns, "rows": rows},
		},
	}
	out, _, err := Sanitize(spec, ReadOnlyNodeTypes)
	if err != nil {
		t.Fatalf("Sanitize returned error: %v", err)
	}
	node := out["items"].([]any)[0].(map[string]any)
	if got := len(node["rows"].([]any)); got != MaxTableRows {
		t.Fatalf("len(rows) = %d, want %d", got, MaxTableRows)
	}
}

// TestSanitize_UnlistedFieldStripped is the core security-property test: a
// node's output must only ever contain fields this package explicitly
// copied and clamped, never anything from the input node it didn't ask for.
func TestSanitize_UnlistedFieldStripped(t *testing.T) {
	spec := map[string]any{
		"items": []any{
			map[string]any{
				"type":    "text",
				"text":    "hi",
				"onclick": "evil()",
				"style":   "position:fixed",
				"href":    "javascript:alert(1)",
			},
		},
	}
	out, _, err := Sanitize(spec, ReadOnlyNodeTypes)
	if err != nil {
		t.Fatalf("Sanitize returned error: %v", err)
	}
	node := out["items"].([]any)[0].(map[string]any)
	for _, unlisted := range []string{"onclick", "style", "href"} {
		if _, present := node[unlisted]; present {
			t.Fatalf("unlisted field %q leaked through: %#v", unlisted, node)
		}
	}
	if len(node) != 2 { // type + text only
		t.Fatalf("node has unexpected fields: %#v", node)
	}
}

// TestSanitize_ExactlyAtNodeCapKeepsAll is the exact-boundary counterpart of
// TestSanitize_OverCapNodeCountTrimmed — MaxNodes flat siblings should all
// survive (this alone would not have caught the container off-by-one below,
// since flat siblings never nest).
func TestSanitize_ExactlyAtNodeCapKeepsAll(t *testing.T) {
	items := make([]any, MaxNodes)
	for i := range items {
		items[i] = map[string]any{"type": "text", "text": "x"}
	}
	_, count, err := Sanitize(map[string]any{"items": items}, ReadOnlyNodeTypes)
	if err != nil {
		t.Fatalf("Sanitize returned error: %v", err)
	}
	if count != MaxNodes {
		t.Fatalf("count = %d, want %d (none should be trimmed exactly at the cap)", count, MaxNodes)
	}
}

// TestSanitize_ContainerNodesReserveOwnBudgetSlot reproduces the review
// finding: a chain of containers must count themselves BEFORE their
// children spend the remaining budget, or the total can overshoot MaxNodes.
// Nests MaxNodes text leaves one level down inside a "row", so a container
// that doesn't reserve its own slot first would let the total reach
// MaxNodes+1.
func TestSanitize_ContainerNodesReserveOwnBudgetSlot(t *testing.T) {
	leaves := make([]any, MaxNodes)
	for i := range leaves {
		leaves[i] = map[string]any{"type": "text", "text": "x"}
	}
	spec := map[string]any{
		"items": []any{
			map[string]any{"type": "row", "children": leaves},
		},
	}
	_, count, err := Sanitize(spec, ReadOnlyNodeTypes)
	if err != nil {
		t.Fatalf("Sanitize returned error: %v", err)
	}
	if count > MaxNodes {
		t.Fatalf("count = %d, exceeds MaxNodes = %d — the row didn't reserve its own budget slot before recursing", count, MaxNodes)
	}
}

// TestSanitize_ExactlyAtDepthKept and TestSanitize_OneOverDepthDropped are
// the exact-boundary counterpart of TestSanitize_OverDepthNodeDropped, which
// only exercised depth 10 against MaxDepth=8.
func TestSanitize_ExactlyAtDepthKept(t *testing.T) {
	var build func(depth int) map[string]any
	build = func(depth int) map[string]any {
		if depth == 0 {
			return map[string]any{"type": "text", "text": "leaf"}
		}
		return map[string]any{"type": "row", "children": []any{build(depth - 1)}}
	}
	// MaxDepth nested rows below the top-level item = depth MaxDepth+1 for
	// the leaf itself (top-level items start at depth 1), matching the
	// convention sanitizeNode already uses.
	spec := map[string]any{"items": []any{build(MaxDepth - 1)}}
	out, _, err := Sanitize(spec, ReadOnlyNodeTypes)
	if err != nil {
		t.Fatalf("Sanitize returned error: %v", err)
	}
	node := out["items"].([]any)[0].(map[string]any)
	depth := 1
	for {
		children, ok := node["children"].([]any)
		if !ok || len(children) == 0 {
			break
		}
		node = children[0].(map[string]any)
		depth++
	}
	if node["type"] != "text" {
		t.Fatalf("leaf at exactly MaxDepth should survive, got %#v at depth %d", node, depth)
	}
}

func TestSanitize_StringExactlyAtCapNotTruncated(t *testing.T) {
	exact := strings.Repeat("a", MaxStringLen)
	spec := map[string]any{
		"items": []any{map[string]any{"type": "text", "text": exact}},
	}
	out, _, err := Sanitize(spec, ReadOnlyNodeTypes)
	if err != nil {
		t.Fatalf("Sanitize returned error: %v", err)
	}
	got := out["items"].([]any)[0].(map[string]any)["text"].(string)
	if got != exact {
		t.Fatalf("a string exactly at MaxStringLen should be untouched; got len %d, want %d", len(got), MaxStringLen)
	}
}

// TestSanitize_TableNonStringNonNumberCellBecomesPlaceholder reproduces the
// review finding: a cell of an unsupported JSON type (bool here) must become
// a placeholder, not be skipped — skipping shifts every later cell in the
// row left by one, misaligning it against the table's "columns" header.
func TestSanitize_TableNonStringNonNumberCellBecomesPlaceholder(t *testing.T) {
	spec := map[string]any{
		"items": []any{
			map[string]any{
				"type":    "table",
				"columns": []any{"Name", "Active", "City"},
				"rows":    []any{[]any{"Alice", true, "NYC"}},
			},
		},
	}
	out, _, err := Sanitize(spec, ReadOnlyNodeTypes)
	if err != nil {
		t.Fatalf("Sanitize returned error: %v", err)
	}
	row := out["items"].([]any)[0].(map[string]any)["rows"].([]any)[0].([]any)
	if len(row) != 3 {
		t.Fatalf("row = %#v, want 3 cells (bool cell replaced, not dropped)", row)
	}
	if row[0] != "Alice" || row[2] != "NYC" {
		t.Fatalf("row = %#v, want column alignment preserved (City still at index 2)", row)
	}
}

func TestClampString_DoesNotSplitMultibyteRune(t *testing.T) {
	// Each CJK character below is 3 bytes in UTF-8; a byte-blind s[:n] at an
	// odd boundary would split one in half and produce invalid UTF-8.
	s := strings.Repeat("你", 10) // 30 bytes
	got := clampString(s, 7)     // cuts mid-character at byte 7 (into the 3rd 你)
	if !strings.HasPrefix(s, got) {
		t.Fatalf("clampString(%q, 7) = %q is not a prefix of the input", s, got)
	}
	for i, r := range got {
		_ = i
		if r == 0xFFFD {
			t.Fatalf("clampString produced a replacement character (invalid UTF-8 split): %q", got)
		}
	}
	// 7 bytes fits 2 complete 你 (6 bytes) — the 3rd is incomplete and must
	// be dropped entirely, not truncated into garbage.
	if got != strings.Repeat("你", 2) {
		t.Fatalf("clampString(%q, 7) = %q, want %q", s, got, strings.Repeat("你", 2))
	}
}

File: web/src/lib/genui/guard.test.ts
import { describe, it, expect } from 'vitest'
import {
  sanitizeSpec,
  READ_ONLY_NODE_TYPES,
  INTERACTIVE_NODE_TYPES,
  MAX_NODES,
  MAX_DEPTH,
  MAX_STRING_LEN,
  MAX_TABLE_ROWS,
  MAX_OPTIONS,
  MAX_TABS,
} from './guard'

// Slice B's inline-fence path is the only path that ever passes the
// interactive types in — render_ui's tool-card path never does (see
// INTERACTIVE_NODE_TYPES's doc comment). Union the two whitelists here to
// mirror fence-split.ts's ALLOWED_TYPES.
const ALL_TYPES: ReadonlySet<string> = new Set([...READ_ONLY_NODE_TYPES, ...INTERACTIVE_NODE_TYPES])

describe('sanitizeSpec: valid input', () => {
  it('passes a well-formed spec through unchanged', () => {
    const spec = {
      title: 'Order status',
      items: [
        { type: 'text', text: 'hello' },
        { type: 'stat', label: 'Revenue', value: '$128,430', tone: 'up' },
      ],
    }
    const { spec: out, count } = sanitizeSpec(spec, READ_ONLY_NODE_TYPES)
    expect(count).toBe(2)
    expect(out?.items).toHaveLength(2)
    expect(out?.items[0]).toEqual({ type: 'text', text: 'hello' })
  })

  it('returns null when items is missing', () => {
    const { spec: out, count } = sanitizeSpec({ title: 'x' }, READ_ONLY_NODE_TYPES)
    expect(out).toBeNull()
    expect(count).toBe(0)
  })

  it('returns null for non-object input (e.g. a still-streaming partial parse)', () => {
    const { spec: out } = sanitizeSpec(null, READ_ONLY_NODE_TYPES)
    expect(out).toBeNull()
  })
})

describe('sanitizeSpec: unknown types', () => {
  it('drops an unknown-type node (and its subtree) without failing the whole spec', () => {
    const spec = {
      items: [
        { type: 'text', text: 'kept' },
        { type: 'iframe', src: 'https://evil.example' },
      ],
    }
    const { spec: out, count } = sanitizeSpec(spec, READ_ONLY_NODE_TYPES)
    expect(count).toBe(1)
    expect(out?.items).toHaveLength(1)
    expect(out?.items[0]).toMatchObject({ type: 'text' })
  })
})

describe('sanitizeSpec: unlisted fields stripped', () => {
  it('strips any field the schema does not know about', () => {
    const spec = {
      items: [{ type: 'text', text: 'hi', onclick: 'evil()', style: 'position:fixed' }],
    }
    const { spec: out } = sanitizeSpec(spec, READ_ONLY_NODE_TYPES)
    const node = out?.items[0] as any
    expect(node.onclick).toBeUndefined()
    expect(node.style).toBeUndefined()
    expect(Object.keys(node).sort()).toEqual(['text', 'type'])
  })
})

describe('sanitizeSpec: node count cap', () => {
  it('trims (does not error) once the total exceeds MAX_NODES', () => {
    const items = Array.from({ length: MAX_NODES + 50 }, () => ({ type: 'text', text: 'x' }))
    const { spec: out, count } = sanitizeSpec({ items }, READ_ONLY_NODE_TYPES)
    expect(count).toBe(MAX_NODES)
    expect(out?.items).toHaveLength(MAX_NODES)
  })

  it('keeps exactly MAX_NODES flat siblings untrimmed', () => {
    const items = Array.from({ length: MAX_NODES }, () => ({ type: 'text', text: 'x' }))
    const { count } = sanitizeSpec({ items }, READ_ONLY_NODE_TYPES)
    expect(count).toBe(MAX_NODES)
  })

  it('reserves a container node its own budget slot before recursing', () => {
    // Same shape as the bug the Go guard review caught: MAX_NODES leaves
    // nested one level down inside a single "row" must not push the total
    // past MAX_NODES.
    const leaves = Array.from({ length: MAX_NODES }, () => ({ type: 'text', text: 'x' }))
    const spec = { items: [{ type: 'row', children: leaves }] }
    const { count } = sanitizeSpec(spec, READ_ONLY_NODE_TYPES)
    expect(count).toBeLessThanOrEqual(MAX_NODES)
  })
})

describe('sanitizeSpec: depth cap', () => {
  function nestedRow(depth: number): any {
    if (depth === 0) return { type: 'text', text: 'leaf' }
    return { type: 'row', children: [nestedRow(depth - 1)] }
  }

  it('drops a leaf past MAX_DEPTH', () => {
    const spec = { items: [nestedRow(10)] } // leaf lands at depth 11
    const { spec: out } = sanitizeSpec(spec, READ_ONLY_NODE_TYPES)
    let node: any = out?.items[0]
    let depth = 1
    while (node.children?.length) {
      node = node.children[0]
      depth++
    }
    expect(node.type).toBe('row') // the leaf itself never survived
    expect(depth).toBeLessThanOrEqual(MAX_DEPTH)
  })

  it('keeps a leaf exactly at MAX_DEPTH', () => {
    const spec = { items: [nestedRow(MAX_DEPTH - 1)] } // leaf lands at depth MAX_DEPTH
    const { spec: out } = sanitizeSpec(spec, READ_ONLY_NODE_TYPES)
    let node: any = out?.items[0]
    while (node.children?.length) node = node.children[0]
    expect(node.type).toBe('text')
  })
})

describe('sanitizeSpec: string clamping', () => {
  it('clamps an over-length string field', () => {
    const long = 'a'.repeat(MAX_STRING_LEN + 100)
    const { spec: out } = sanitizeSpec({ items: [{ type: 'text', text: long }] }, READ_ONLY_NODE_TYPES)
    expect((out?.items[0] as any).text).toHaveLength(MAX_STRING_LEN)
  })

  it('leaves a string exactly at the cap untouched', () => {
    const exact = 'a'.repeat(MAX_STRING_LEN)
    const { spec: out } = sanitizeSpec({ items: [{ type: 'text', text: exact }] }, READ_ONLY_NODE_TYPES)
    expect((out?.items[0] as any).text).toBe(exact)
  })

  it('does not split a surrogate pair (emoji) when clamping', () => {
    const emoji = '🎉'.repeat(MAX_STRING_LEN) // each 🎉 is 2 UTF-16 code units
    const { spec: out } = sanitizeSpec({ items: [{ type: 'text', text: emoji }] }, READ_ONLY_NODE_TYPES)
    const text = (out?.items[0] as any).text as string
    // A valid clamp never leaves a lone leading low surrogate.
    expect(text.charCodeAt(0)).not.toBeGreaterThanOrEqual(0xdc00)
    expect(Array.from(text).every((ch) => ch === '🎉')).toBe(true)
  })
})

describe('sanitizeSpec: enum fields', () => {
  it('drops an invalid tone rather than passing it through', () => {
    const { spec: out } = sanitizeSpec({ items: [{ type: 'badge', text: 'x', tone: 'rainbow' }] }, READ_ONLY_NODE_TYPES)
    expect((out?.items[0] as any).tone).toBeUndefined()
  })

  it('keeps a valid tone', () => {
    const { spec: out } = sanitizeSpec({ items: [{ type: 'badge', text: 'x', tone: 'success' }] }, READ_ONLY_NODE_TYPES)
    expect((out?.items[0] as any).tone).toBe('success')
  })
})

describe('sanitizeSpec: progress value clamping', () => {
  it('clamps out-of-range values to [0, 100]', () => {
    const { spec: out } = sanitizeSpec(
      { items: [{ type: 'progress', value: 150 }, { type: 'progress', value: -20 }] },
      READ_ONLY_NODE_TYPES
    )
    expect((out?.items[0] as any).value).toBe(100)
    expect((out?.items[1] as any).value).toBe(0)
  })
})

describe('sanitizeSpec: table caps', () => {
  it('caps rows at MAX_TABLE_ROWS', () => {
    const rows = Array.from({ length: MAX_TABLE_ROWS + 20 }, () => ['x', 'y'])
    const { spec: out } = sanitizeSpec({ items: [{ type: 'table', columns: ['a', 'b'], rows }] }, READ_ONLY_NODE_TYPES)
    expect((out?.items[0] as any).rows).toHaveLength(MAX_TABLE_ROWS)
  })

  it('replaces a non-string/number cell with a placeholder to preserve column alignment', () => {
    const spec = { items: [{ type: 'table', columns: ['Name', 'Active', 'City'], rows: [['Alice', true, 'NYC']] }] }
    const { spec: out } = sanitizeSpec(spec, READ_ONLY_NODE_TYPES)
    const row = (out?.items[0] as any).rows[0]
    expect(row).toHaveLength(3)
    expect(row[0]).toBe('Alice')
    expect(row[2]).toBe('NYC')
  })
})

describe('sanitizeSpec: interactive nodes are rejected on the read-only whitelist', () => {
  it('render_ui tool-card path (READ_ONLY_NODE_TYPES) drops a button node', () => {
    const spec = { items: [{ type: 'button', label: 'Go', action: 'go' }] }
    const { spec: out, count } = sanitizeSpec(spec, READ_ONLY_NODE_TYPES)
    expect(count).toBe(0)
    expect(out?.items).toHaveLength(0)
  })
})

describe('sanitizeSpec: button node', () => {
  it('passes a valid button through with fields clamped', () => {
    const spec = { items: [{ type: 'button', label: 'Refresh', action: 'refresh', variant: 'primary', payload: { range: '7d' } }] }
    const { spec: out, count } = sanitizeSpec(spec, ALL_TYPES)
    expect(count).toBe(1)
    expect(out?.items[0]).toEqual({
      type: 'button',
      label: 'Refresh',
      action: 'refresh',
      variant: 'primary',
      payload: { range: '7d' },
    })
  })

  it('drops an invalid variant rather than passing it through', () => {
    const spec = { items: [{ type: 'button', label: 'Go', action: 'go', variant: 'rainbow' }] }
    const { spec: out } = sanitizeSpec(spec, ALL_TYPES)
    expect((out?.items[0] as any).variant).toBeUndefined()
  })

  it('clamps over-length label/action strings', () => {
    const long = 'a'.repeat(MAX_STRING_LEN + 50)
    const spec = { items: [{ type: 'button', label: long, action: long }] }
    const { spec: out } = sanitizeSpec(spec, ALL_TYPES)
    const node = out?.items[0] as any
    expect(node.label).toHaveLength(MAX_STRING_LEN)
    expect(node.action).toHaveLength(MAX_STRING_LEN)
  })

  it('truncates a payload nested past MAX_DEPTH to an empty object', () => {
    let payload: any = { leaf: 'x' }
    for (let i = 0; i < MAX_DEPTH + 3; i++) payload = { nested: payload }
    const spec = { items: [{ type: 'button', label: 'Go', action: 'go', payload }] }
    const { spec: out } = sanitizeSpec(spec, ALL_TYPES)
    const node = out?.items[0] as any
    // Walk down the surviving payload until a level truncates to {}.
    let level = node.payload
    let depth = 1
    while (level && level.nested !== undefined && Object.keys(level).length > 0) {
      level = level.nested
      depth++
    }
    // sanitizePayload(obj, depth=1) truncates once depth > MAX_DEPTH, i.e.
    // at the call for depth MAX_DEPTH+1 — one further in than the node-tree
    // MAX_DEPTH convention (which counts a top-level item as depth 1 and
    // truncates children past MAX_DEPTH itself). Different function, its
    // own off-by-one convention — asserted here rather than assumed.
    expect(depth).toBeLessThanOrEqual(MAX_DEPTH + 1)
    expect(level).toEqual({})
  })

  it('caps a payload object with more than MAX_OPTIONS keys at one level', () => {
    const wide: Record<string, string> = {}
    for (let i = 0; i < MAX_OPTIONS + 20; i++) wide[`k${i}`] = 'v'
    const spec = { items: [{ type: 'button', label: 'Go', action: 'go', payload: wide }] }
    const { spec: out } = sanitizeSpec(spec, ALL_TYPES)
    const node = out?.items[0] as any
    expect(Object.keys(node.payload)).toHaveLength(MAX_OPTIONS)
  })

  it('caps a payload array with more than MAX_OPTIONS entries', () => {
    const spec = {
      items: [{ type: 'button', label: 'Go', action: 'go', payload: { list: Array.from({ length: MAX_OPTIONS + 20 }, (_, i) => i) } }],
    }
    const { spec: out } = sanitizeSpec(spec, ALL_TYPES)
    const node = out?.items[0] as any
    expect(node.payload.list).toHaveLength(MAX_OPTIONS)
  })
})

describe('sanitizeSpec: input node', () => {
  it('passes a valid input through with fields clamped', () => {
    const spec = { items: [{ type: 'input', field: 'name', label: 'Name', placeholder: 'Enter name', value: 'Alice' }] }
    const { spec: out } = sanitizeSpec(spec, ALL_TYPES)
    expect(out?.items[0]).toEqual({ type: 'input', field: 'name', label: 'Name', placeholder: 'Enter name', value: 'Alice' })
  })

  it('silently drops an inputType field even if the model sends one', () => {
    const spec = { items: [{ type: 'input', field: 'password', inputType: 'password' }] }
    const { spec: out } = sanitizeSpec(spec, ALL_TYPES)
    const node = out?.items[0] as any
    expect(node.inputType).toBeUndefined()
    expect(Object.keys(node).sort()).toEqual(['field', 'type'])
  })
})

describe('sanitizeSpec: select node', () => {
  it('passes a valid select through with options clamped', () => {
    const spec = {
      items: [
        {
          type: 'select',
          field: 'range',
          label: 'Range',
          value: '7d',
          options: [{ label: '7 days', value: '7d' }, { label: '30 days', value: '30d' }],
        },
      ],
    }
    const { spec: out } = sanitizeSpec(spec, ALL_TYPES)
    expect(out?.items[0]).toEqual({
      type: 'select',
      field: 'range',
      label: 'Range',
      value: '7d',
      options: [{ label: '7 days', value: '7d' }, { label: '30 days', value: '30d' }],
    })
  })

  it('caps options at MAX_OPTIONS', () => {
    const options = Array.from({ length: MAX_OPTIONS + 10 }, (_, i) => ({ label: `o${i}`, value: `${i}` }))
    const spec = { items: [{ type: 'select', field: 'x', options }] }
    const { spec: out } = sanitizeSpec(spec, ALL_TYPES)
    expect((out?.items[0] as any).options).toHaveLength(MAX_OPTIONS)
  })
})

describe('sanitizeSpec: checkbox / switch nodes', () => {
  it('passes a valid checkbox through', () => {
    const spec = { items: [{ type: 'checkbox', field: 'agree', label: 'I agree', checked: true }] }
    const { spec: out } = sanitizeSpec(spec, ALL_TYPES)
    expect(out?.items[0]).toEqual({ type: 'checkbox', field: 'agree', label: 'I agree', checked: true })
  })

  it('passes a valid switch through', () => {
    const spec = { items: [{ type: 'switch', field: 'notify', checked: false }] }
    const { spec: out } = sanitizeSpec(spec, ALL_TYPES)
    expect(out?.items[0]).toEqual({ type: 'switch', field: 'notify', checked: false })
  })

  it('drops a non-boolean checked value', () => {
    const spec = { items: [{ type: 'checkbox', field: 'agree', checked: 'yes' }] }
    const { spec: out } = sanitizeSpec(spec, ALL_TYPES)
    expect((out?.items[0] as any).checked).toBeUndefined()
  })
})

describe('sanitizeSpec: radio node', () => {
  it('passes a valid radio through with options clamped', () => {
    const spec = {
      items: [
        {
          type: 'radio',
          field: 'plan',
          value: 'pro',
          options: [{ label: 'Free', value: 'free' }, { label: 'Pro', value: 'pro' }],
        },
      ],
    }
    const { spec: out } = sanitizeSpec(spec, ALL_TYPES)
    expect(out?.items[0]).toEqual({
      type: 'radio',
      field: 'plan',
      value: 'pro',
      options: [{ label: 'Free', value: 'free' }, { label: 'Pro', value: 'pro' }],
    })
  })
})

describe('sanitizeSpec: tabs node', () => {
  it('passes valid tabs through with children sanitized', () => {
    const spec = {
      items: [
        {
          type: 'tabs',
          tabs: [
            { label: 'Overview', children: [{ type: 'text', text: 'hi' }] },
            { label: 'Details', children: [{ type: 'stat', label: 'Revenue', value: '$1' }] },
          ],
        },
      ],
    }
    const { spec: out } = sanitizeSpec(spec, ALL_TYPES)
    const node = out?.items[0] as any
    expect(node.tabs).toHaveLength(2)
    expect(node.tabs[0]).toEqual({ label: 'Overview', children: [{ type: 'text', text: 'hi' }] })
  })

  it('caps tabs at MAX_TABS', () => {
    const tabs = Array.from({ length: MAX_TABS + 3 }, (_, i) => ({ label: `t${i}`, children: [] }))
    const spec = { items: [{ type: 'tabs', tabs }] }
    const { spec: out } = sanitizeSpec(spec, ALL_TYPES)
    expect((out?.items[0] as any).tabs).toHaveLength(MAX_TABS)
  })
})
